packages/server/src/transform/mpegts.ts

```diff
@@ -0,0 +1,185 @@
+import { nonstandard as WRTCNonstandard } from "wrtc";
+import { PassThrough } from "stream";
+import ffmpeg from "fluent-ffmpeg";
+
+import { createServer } from "net";
+import path from "path";
+import { fstat, statSync, unlink, unlinkSync } from "fs";
+
+const { RTCVideoSink, RTCAudioSink } = WRTCNonstandard;
+
+class UnixStream {
+  private locator: string;
+
+  constructor(stream, id, onSocket) {
+    const socketPath = path.resolve("/tmp/", id + ".sock");
+    this.locator = "unix:" + socketPath;
+
+    try {
+      statSync(socketPath);
+      unlinkSync(socketPath);
+    } catch (err) {}
+    let server;
+    try {
+      server = createServer(onSocket);
+    } catch (err) {
+      console.log(err);
+    }
+    stream.on("finish", () => {
+      server.close();
+    });
+    stream.on("error", (err) => {
+      console.log(err.message);
+    });
+    server.listen(socketPath);
+  }
+
+  get url() {
+    return this.locator;
+  }
+}
+
+function StreamInput(stream, id) {
+  return new UnixStream(stream, id, (socket) => {
+    stream.on("error", (err) => {
+      console.log(`INPUT(${id}):` + err.message);
+    });
+    socket.on("error", (err) => {
+      console.log(`OUTPUT(${id}):` + err.message);
+    })
+    stream.pipe(socket);
+  });
+}
+
+var MPEGTS_INTER_PORT = 9000;
+
+export class MPEGTS {
+  private pc: RTCPeerConnection;
+  private interPort: number;
+  private streams;
+  private id: string;
+
+  constructor(pc: RTCPeerConnection, id) {
+    this.pc = pc;
+    this.interPort = MPEGTS_INTER_PORT++;
+    this.id = id;
+
+    let videoSink;
+    let audioSink;
+
+    this.pc.getReceivers().forEach(({ track }) => {
+      if (track.kind === "video") {
+        videoSink = new RTCVideoSink(track);
+      } else if(track.kind === "audio") {
+        audioSink = new RTCAudioSink(track);
+      }
+    });
+
+    this.streams = [];
+
+    videoSink.addEventListener("frame", ({ frame: { width, height, data }}) => {
+      const streamId = width + "x" + height;
+      if (!this.streams[0] || (this.streams[0] && this.streams[0].streamId !== streamId)) {
+        // A new frame size has arrived we need to create a new stream and stop
+        // the others.
+        const stream = this.createStream(streamId, audioSink);
+        this.streams.unshift(stream);
+        this.streams.forEach(item => {
```

Comment:
why are we not ending all the streams first and then adding the new one? ( Also, shouldn't we remove the stopped streams from the array? )

Reply:
Fair point. Could try that. My reasoning was that it would cause an interruption in the mpegts output but now of course we risk two processes writing at the same time

Reply:
Ending first caused an unrecoverable interruption, but yes we could remove the stopped streams from the array.